### PR TITLE
Refactored hydration of job entities

### PR DESCRIPTION
**Pull request by fedys**
refs gh-256

### CODE REVIEW

**Comment by fedys** [comment]
There are still some missing unit tests. I will continue with facet implementation after I complete the unit tests.

**Comment by fedys** [comment]
@TiSiE great!

**Comment by fedys** [comment]
Do you want me to implement filtering by clicking on region facets? If yes, what should happen on clicking a facet region link?

**Comment by TiSiE** [comment]
I copy the mail from @cbleek here for completeness:

> yes, please implement it.
> 
> https://www.stepstone.de/ has made it quite good. If you click on a region, a new search is done. The query is extended by "AND region=<selected region>' and the filtered Region is shown. So the user can remove the filter.
> 
> if you click more than one region, the query is extended by AND (region=<selected region1> OR region=<selected region2>)
> 
> Doing the "OR" Search in Solr can be done by: regionList:(Bayern OR Hessen)
> 
> https://yawik.org:8443/solr/YawikJobs/select?fq=regionList:(Bayern%20OR%20Hessen)&indent=on&q=*:*&wt=json
> 
>  

-----
Maybe we should normalize the facets entity. So switching to another module providing similar features is easy (e.g. ElasticSearch Backend)

**Comment by fedys** [comment]
@TiSiE I would wait with facets normalization until we finish the current Solr implementation and start implementing another backend providing similar functionality

**Comment by TiSiE** [comment]
@fedys You're right, sometimes I am a little overeager...

**Comment by fedys** [comment]
I will complete unit tests as soon as you tell me you are happy with the current implementation of the facet filtering.

**Comment by fedys** [comment]
I noticed that the behavior of facet counts is not proper if an user checks any facet checkbox. A [facet tag excluding](http://yonik.com/multi-select-faceting/)  has to be applied to fix it. Do you want me to apply the fix?

File: module/Solr/src/Solr/Bridge/ResultConverter.php [hunk 1]
```diff
@@ -80,12 +81,20 @@ public function convert(AbstractPaginationQuery $filter, ArrayAccess $response)
...
         foreach ($response['response']['docs'] as $doc) {
             // check if entity exists
-            if (!isset($entities[$doc->id])) {
-                // skip non-existent entity
-                continue;
+            if (isset($entities[$doc->id])) {
+                // use found entity
+                $entity = $entities[$doc->id];
+            } else {
+                if (!isset($emptyEntity)) {
```
Comment (TiSiE) [inline module/Solr/src/Solr/Bridge/ResultConverter.php]: Better use dedicated dummy entities. If someone messed up one, it won't affect the others.

**Comment by fedys** [comment]
@TiSiE I cannot reproduce it in my environment (PHP 5.6.26-1+deb.sury.org~trusty+1). What PHP version do you have? What Solr source do you use?

**Comment by fedys** [comment]
@TiSiE Maybe it can be related to [this one](https://github.com/phpnode/YiiSolr/blob/master/ASolrQueryResponse.php#L162)

**Comment by TiSiE** [comment]
@fedys yes, this is exactly the error. I'm using PHP 5.5.9-1ubuntu4.17and pecl-solr 2.4.0

**Comment by fedys** [comment]
@TiSiE okay, I will fix it.

**Comment by TiSiE** [comment]
Probably this issue: https://bugs.php.net/bug.php?id=67300&edit=1

Cast an array with empty keys to an object DOES NOT create an error, but trying to access the empty key (when iterating, eg) produce the Notice.

I can reproduce this behaviour in my PHP:
```
php > $a = ['test' => 'ok', '' => 'error'];
php > $b = (object) $a;
php > 
php > 
php > echo $b->test;
okphp > 
php > 
php > var_dump($b);
PHP Notice:  Illegal member variable name in php shell code on line 1
PHP Stack trace:
PHP   1. {main}() php shell code:0
PHP   2. var_dump(...) php shell code:1
class stdClass#1 (2) {
  public $test =>
  string(2) "ok"
  public $ =>
  string(5) "error"
}
php > 
php > 
php > foreach ($b as $k => $v) { echo "$k: $v\n"; }
test: ok
PHP Notice:  Illegal member variable name in php shell code on line 1
PHP Stack trace:
PHP   1. {main}() php shell code:0
php > 


```

**Comment by fedys** [comment]
@TiSiE great! I am going to complete unit tests.

**Pull request merged**